gdal/frmts/grib/gribdataset.cpp

```diff
@@ -981,10 +1086,137 @@ void GRIBRasterBand::UncacheData()
 
 GRIBRasterBand::~GRIBRasterBand()
 {
-    CPLFree(longFstLevel);
+    if (longFstLevel != nullptr)
+        CPLFree(longFstLevel);
     UncacheData();
 }
 
+
+/************************************************************************/
+/*                           InventoryWrapperGrib                       */
+/************************************************************************/
+class InventoryWrapperGrib : public gdal::grib::InventoryWrapper
+{
+  public:
+    explicit InventoryWrapperGrib(VSILFILE *fp) : gdal::grib::InventoryWrapper()
+    {
+        result_ = GRIB2Inventory(fp, &inv_, &inv_len_, 0 /* all messages */, &num_messages_);
+    }
+
+    ~InventoryWrapperGrib() override
+    {
+        if (inv_ == nullptr)
+            return;
+        for (uInt4 i = 0; i < inv_len_; i++)
+        {
+            GRIB2InventoryFree(inv_ + i);
+        }
+        free(inv_);
+    }
+};
+
+/************************************************************************/
+/*                           InventoryWrapperSidecar                    */
+/************************************************************************/
+
+class InventoryWrapperSidecar : public gdal::grib::InventoryWrapper
+{
+  public:
+    explicit InventoryWrapperSidecar(VSILFILE *fp) : gdal::grib::InventoryWrapper()
+    {
+        VSIFSeekL(fp, 0, SEEK_END);
+        vsi_l_offset length = VSIFTellL(fp);
+        char *pszSidecar = new char[length + 1];
+        VSIFSeekL(fp, 0, SEEK_SET);
+        if (VSIFReadL(pszSidecar, static_cast<size_t>(length), 1, fp) != 1)
+        {
+            result_ = -1;
+            return;
+        }
+        pszSidecar[length] = 0;
+
+        char **papszMsgs = CSLTokenizeString2(pszSidecar, "\n", CSLT_PRESERVEQUOTES);
```

Comment:
to avoid manual memory feeing:
```suggestion
        CPLStringList aosMsgs(CSLTokenizeString2(pszSidecar, "\n", CSLT_PRESERVEQUOTES));
```